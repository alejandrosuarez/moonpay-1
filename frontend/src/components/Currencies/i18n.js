const translations = {
  fr: {
    textResults: "RÃ©sultats",
  },
  en: {
    textResults: "Results",
  },
};

export default translations;
