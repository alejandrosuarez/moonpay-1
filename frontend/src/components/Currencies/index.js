import React, { useEffect, useState } from "react";
import Currency from "../Currency";
import * as API from "../../services/api";
import { ApiRoutes } from "../../settings/routes";
import "./index.scss";
import translations from "./i18n";
import { SORT_BY_RANDOM } from "../../settings/constants";
import { useSelector } from "react-redux";

const Currencies = () => {
  // Get the filter state from redux state
  const filter = useSelector((state) => state.filter);
  const { locale } = useSelector((state) => state.locale);

  // Arrow function States
  const [currencies, setCurrencies] = useState([]);
  const [filteredCurrencies, setFilteredCurrencies] = useState([]);

  /**
   * Use Effect when the component is mounted
   */
  useEffect(() => {
    getCurrencies();
  }, []);

  /**
   * Use Effect when the filter state is updated
   */
  useEffect(() => {
    handleCurrencies(currencies);
  }, [filter]);

  /**
   * Method to get currencies
   */
  const getCurrencies = () => {
    // Call API to get currencies
    API.get(ApiRoutes.CURRENCIES)
      .then((data) => {
        setCurrencies([...data]);
        handleCurrencies([...data]);
      })
      .catch((error) => {
        // Init states
        setCurrencies([]);
      });
  };

  /**
   * Method to apply filters and sort in the currencies list
   * @param data
   */
  const handleCurrencies = (data) => {
    let items = filterCurrencies(data);
    setFilteredCurrencies(sortCurrencies(items));
  };

  /**
   * Method to filter
   */
  const filterCurrencies = (items) => {
    return items.filter(
      (currency) =>
        (filter.isSupportsTestMode ? currency.supportsTestMode : true) &&
        (filter.isSupportedInUS ? currency.isSupportedInUS : true)
    );
  };

  /**
   * Method to sort by name, code or to shuffle currencies
   */
  const sortCurrencies = (items) => {
    let sortedCurrencies = [];

    if (filter.sortBy === SORT_BY_RANDOM) {
      // Shuffle the list
      sortedCurrencies = items.sort(() => 0.5 - Math.random());
    } else {
      // Sort by name or by code the list
      sortedCurrencies = items.sort((currencyFirst, currencySecond) => {
        if (currencyFirst[filter.sortBy] < currencySecond[filter.sortBy]) {
          return -1;
        }
        if (currencyFirst[filter.sortBy] > currencySecond[filter.sortBy]) {
          return 1;
        }
        return 0;
      });
    }

    return sortedCurrencies;
  };

  return (
    <>
      <h1>
        {filteredCurrencies.length} {translations[locale].textResults}
      </h1>
      <ul className="m-currencies">
        {filteredCurrencies.map((currency, index) => (
          <li key={index} className="m-currencies__item">
            <Currency currency={currency} sortBy={filter.sortBy} />
          </li>
        ))}
      </ul>
    </>
  );
};

export default Currencies;
