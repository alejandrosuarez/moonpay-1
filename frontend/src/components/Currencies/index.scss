@import "../../styles/index.scss";

.m-currencies {
  display: grid;
  gap: rem(15) calc(16.67% - 160px);
  width: 100%;
  grid-template-columns: repeat(3, minmax(0px, 1fr));

  &__item {
    @include card(true);
    margin: rem(10);
    text-align: center;
  }
}

@media only screen and (max-width: 478px) {
  .m-currencies {
    grid-template-columns: repeat(1, minmax(0px, 1fr));
  }
}

@media only screen and (min-width: 479px) and (max-width: 768px) {
  .m-currencies {
    grid-template-columns: repeat(2, minmax(0px, 1fr));
  }
}
