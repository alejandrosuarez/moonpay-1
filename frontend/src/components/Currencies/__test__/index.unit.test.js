import React from "react";
import ReactDom from "react-dom";
import Currencies from "../index";
import { cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { create } from "react-test-renderer";

jest.mock("react-redux", () => ({
  useSelector: () => ({ locale: "en" }),
}));

describe("Currencies Component", () => {
  afterEach(cleanup);

  it("Renders without crashing", () => {
    const div = document.createElement("div");
    ReactDom.render(<Currencies />, div);
  });

  it("Matches snapshot empty", () => {
    const app = create(<Currencies />).toJSON();
    expect(app).toMatchSnapshot();
  });
});
