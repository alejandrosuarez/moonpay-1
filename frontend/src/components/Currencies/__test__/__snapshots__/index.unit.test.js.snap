// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Currencies Component Matches snapshot empty 1`] = `
Array [
  <h1>
    0
     
    Results
  </h1>,
  <ul
    className="m-currencies"
  />,
]
`;
